lib/SILGen/SILGenPoly.cpp

```diff
@@ -1876,6 +1883,29 @@ void ResultPlanner::plan(AbstractionPattern innerOrigType,
                          AbstractionPattern outerOrigType,
                          CanType outerSubstType,
                          PlanData &planData) {
+  // Conversion from `() -> T` to `() -> Void` is allowed when
+  // the argument is a closure.
+  if (!innerSubstType->isVoid() && outerSubstType->isVoid()) {
+    auto ignoreNextResult = [&]() {
+      auto innerResult = claimNextInnerResult(planData);
+      if (innerResult.isFormalIndirect() &&
+          SGF.silConv.isSILIndirect(innerResult))
+        (void)addInnerIndirectResultTemporary(planData, innerResult);
+      else
+        addIgnoreDirectResult();
+    };
+
+    if (innerOrigType.isTuple()) {
+      auto innerTuple = cast<TupleType>(innerSubstType);
+      for (unsigned i = 0, n = innerTuple->getNumElements(); i != n; ++i)
+        ignoreNextResult();
```

Comment:
1. This needs be recursive — we expand tuples recursively, so this needs to do it, too.
2. This needs to follow the inner abstraction pattern, not the inner substituted type.  There could be a generic type that's been substituted to a tuple.

I would suggest breaking this out as a method on the planner.